ADR/ADR030-use-of-aws-lambda.md

```diff
@@ -0,0 +1,73 @@
+# ADR030: Use AWS Lambda small tasks run infrequently
+
+Date: 2024-05-17
+
+## Status
+
+> Accepted or Rejected
+>
+> Valid ADR statuses are: "Proposed", "Accepted", "Rejected", "Deprecated", "Superseded by ADRzzz", but since we use the GitHub Pull request mechanism to approve ADRs, "Proposed" isn't applicable.
+
+## Context
+
+There are, in this context, two types of code we want to consider
+
+1. Application code
+2. Glue code
+
+Application code is written by the developers on the team, and makes up the GOV.UK Forms service and all of
+its functionality. It is executed to handle HTTP requests received by a long-lived web server process, and it has access
+to stateful data storage like PostgreSQL and Redis.
+
+In contrast, glue code isn't part of any application, but is required to support  the function of the service or its 
+infrastructure in some way. Examples of that could include code for integrating two  independent systems, or code for 
+testing the state of something periodically.
+
+This ADR addresses where we run infrastructure code, and provides guidance on when not to follow the ADR.
+
+## Decision: use AWS Lambda
+
+Our application code runs in containers on [AWS ECS on Fargate](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/AWS_Fargate.html).
+It is a managed service which is well suited to running long-running processes like web servers. It has a somewhat involved 
+setup process and its use of containers incurs some maintenance and operational overhead, but that is a worthwhile 
+trade-off for its functionality and stability. It is also able to run short-lived tasks.
+
+However, for infrastructure code, which is often characterized by being small, quick to run, and not requiring a connection to
+any stateful data stores, the overhead of the different pipelines and configurations required to run in ECS becomes 
+disproportionate to the size of the code and the frequency with which we think about it. We can quickly end up with more
+code for running and configuring the thing than code in the thing we want to run. 
+
+AWS Lambda is AWS' serverless functions-as-a-service service. It be given a source code bundle and it can execute the code within 
+based on a small amount of configuration. Its primary target use case is in running code in response to events, without the
+need for a long running server. This matches up well with the types of use cases we have for glue code.
+
+We should use AWS Lambda for workloads that meet these criteria:
```

Comment:
Can we add 'written in Ruby' to this please to avoid further divergence of languages within the team and the temptation to use other languages supported by lambda.

Reply:
Added